crates/sui-types/src/base_types.rs

```diff
@@ -94,20 +105,178 @@ pub fn random_object_ref() -> ObjectRef {
     )
 }
 
+/// Wrapper around StructTag with a space-efficient representation for common types like coins
+/// The StructTag for a gas coin is 84 bytes, so using 1 byte instead is a win
+#[derive(Eq, PartialEq, PartialOrd, Ord, Debug, Clone, Deserialize, Serialize, Hash)]
+pub enum MoveType {
+    /// A SUI coin (i.e., 0x2::coin::Coin<0x2::sui::SUI>)
+    GasCoin,
+    /// A record of a staked SUI coin (i.e., 0x2::staking_pool::StakedSui)
+    StakedSui,
+    /// A non-SUI coin type (i.e., 0x2::coin::Coin<inner>)
+    Coin { inner: TypeTag },
```

Comment:
nit, maybe `coin_type` or something? What do we call in Move? Probably just `T` lol
But maybe then just `Coin(TypeTag)`?